### PR TITLE
Limiting the processing of a block in Runtime by delaying receipts

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
- Passing `gas_limit` from chunk_extra to the Runtime.
- Processing receipts until the current total burnt_gas doesn't reach or exceed the `gas_limit`.
- Update balance checker to work with the delayed receipts.
- Tests to cover this. Update challenge test partial state, cause we touch another key in the Runtime to check for delayed receipts.

Going to update nomicon before merging this PR.

Ref #1507

### CODE REVIEW

File: runtime/runtime/src/lib.rs [hunk 1]
```diff
@@ -948,29 +960,80 @@ impl Runtime {
...
-            .for_each(|outcome_with_id| outcomes.push(outcome_with_id));
+            .for_each(|outcome_with_id| {
+                *total_gas_burnt += outcome_with_id.outcome.gas_burnt;
+                outcomes.push(outcome_with_id);
+            });
+            Ok(())
+        };
+
+        while delayed_receipts_indices.first_index < delayed_receipts_indices.next_available_index {
+            if total_gas_burnt >= apply_state.gas_limit {
```
Comment (nearmax) [inline runtime/runtime/src/lib.rs]: It seems like with this logic and `if total_gas_burnt < apply_state.gas_limit {` logic below `gas_limit` can be exceeded, (and will be exceeded in most cases) `test_apply_delayed_receipts` seems to demonstrate it too. Since `test_apply_delayed_receipts` assigns 1 gas_limit and we exceed it with every `apply` but still process one receipt on each step it takes exactly n steps to process all the receipts.

I think it is safer if our logic actually prevents gas_limit to be exceeded.

File: runtime/runtime/src/lib.rs [hunk 2]
```diff
@@ -1245,9 +1308,60 @@ mod tests {
...
                 &apply_state,
-                &[Receipt::new_refund(&account_id, small_refund)],
+                &[Receipt::new_refund(&alice_account(), small_refund)],
                 &[],
             )
             .unwrap();
     }
+
+    #[test]
+    fn test_apply_delayed_receipts() {
```
Comment (nearmax) [inline runtime/runtime/src/lib.rs]: Would be useful to have tests for corner cases, like: 
1) we have local and new receipts but there are so many delayed receipts that we only process those;
2) we run out of delayed receipts and continue processing receipts normally (test_apply_delayed_receipts stops right after the delayed receipts are all processed)

File: chain/client/src/client.rs [hunk 3]
```diff
@@ -1042,18 +1043,20 @@ impl Client {
...
+                        // Not being able to fetch a state root most likely implies that we haven't
+                        //     caught up with the next epoch yet.
+                        return self.forward_tx(tx);
+                    }
+                };
             match self.runtime_adapter.validate_tx(
                 head.height + 1,
                 cur_block_header.inner.timestamp,
                 gas_price,
+                gas_limit,
```
[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply chain/client/src/client.rs]: It actually doesn't use `gas_limit` during validate_tx, but it uses `ApplyState`. I either should not put it into `ApplyState` or just put `gas_limit: u64::max_value()` in validate_tx

File: runtime/runtime/src/lib.rs (hunk 1 above)
Comment (evgenykuzyakov) [reply runtime/runtime/src/lib.rs]: Firstly, the Runtime might stuck with not receipt being produced due to large receipt that would exceed the gas_limit.

We can't estimate the gas_burnt for a receipt before processing it. If we already processed this receipt, there are no reason to not include it into the block.
Challenges would follow the same logic and always produce the same amount of receipts.

File: chain/client/src/client.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply chain/client/src/client.rs]: using `gas_limit: u64::max_value()`

File: runtime/runtime/src/lib.rs (hunk 2 above)
Comment (evgenykuzyakov) [reply runtime/runtime/src/lib.rs]: added

File: chain/client/src/client.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (nearmax) [reply chain/client/src/client.rs]: Just make it `Option<Gas>` in `ApplyState`, with `u64::max_value()` we are using an implicit `None`. It is better to make it explicit for readability reasons.

[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply chain/client/src/client.rs]: Good idea

**Comment by evgenykuzyakov** [comment]
Documentation for delayed receipts and Runtime crate: https://github.com/nearprotocol/nomicon/pull/7

File: test-utils/testlib/src/fees_utils.rs [hunk 4]
```diff
@@ -4,7 +4,7 @@ use near::config::INITIAL_GAS_PRICE;
 use near_primitives::types::{Balance, Gas};
 use near_runtime_fees::RuntimeFeesConfig;
 
-const GAS_PRICE: u128 = INITIAL_GAS_PRICE;
+pub const GAS_PRICE: u128 = INITIAL_GAS_PRICE;
```
Comment (ilblackdragon) [inline test-utils/testlib/src/fees_utils.rs]: why do we need this constant if it's just one from near::config?

Comment (evgenykuzyakov) [reply test-utils/testlib/src/fees_utils.rs]: Runtime didn't import config directly, but use testlib fees. I can import near::config as well.

File: runtime/runtime/src/lib.rs [hunk 5]
```diff
@@ -1153,13 +1219,17 @@ impl Runtime {
 
 #[cfg(test)]
 mod tests {
+    use super::*;
+
+    use near::config::INITIAL_GAS_PRICE;
```
Comment (ilblackdragon) [inline runtime/runtime/src/lib.rs]: this is a bit weird to depend on near for just a constant that doesn't matter anyway.

File: near/src/runtime.rs [hunk 6]
```diff
@@ -635,6 +637,8 @@ impl RuntimeAdapter for NightshadeRuntime {
             epoch_length: self.genesis_config.epoch_length,
             gas_price,
             block_timestamp,
+            // NOTE: verify transaction doesn't use gas limit
+            gas_limit: None,
```
Comment (ilblackdragon) [inline near/src/runtime.rs]: should we use gas limit for verification as well?
because if block limit is small-ish and tx is gigantic, it won't fit in the block as well?

Comment (evgenykuzyakov) [reply near/src/runtime.rs]: We probably shouldn't have TXs that are large enough to not fit into the block. We'll be able to refactor this later.

**Comment by evgenykuzyakov** [comment]
> On a second look, I think we should split gas_limit into transaction_gas_limit and receipt_gas_limit, instead of dividing gas_limit by 2.

This is not an easy fix, cause it requires to change half of the chain, cause we recalculate `gas_limit` in the block limit.
Let's do this in a different PR.

File: runtime/runtime/src/lib.rs (hunk 5 above)
Comment (evgenykuzyakov) [reply runtime/runtime/src/lib.rs]: It was used within `gas_burnt_to_reward`. So right now it's indirect dependency. I can refactor all of this in a different PR to remove dependency on `near` from `fees_utils`

**Comment by nearmax** [comment]
https://github.com/nearprotocol/nearcore/issues/1700

File: runtime/runtime/src/lib.rs (hunk 5 above)
Comment (evgenykuzyakov) [reply runtime/runtime/src/lib.rs]: #1701
